graveler/committed/tree/writer.go

```diff
@@ -0,0 +1,180 @@
+package tree
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"hash/fnv"
+
+	"github.com/treeverse/lakefs/graveler"
+	"github.com/treeverse/lakefs/graveler/committed"
+)
+
+type writer struct {
+	manager    committed.PartManager
+	partWriter committed.Writer // writer for the current part
+	partPath   committed.Namespace
+	lastKey    committed.Key
+	treeWriter committed.Writer // writer for the tree representation
+	treePath   committed.Namespace
+	breakRatio int // indicates when to break the parts
+}
+
+var ErrUnsorted = errors.New("record should be bigger then all records previously written to writer")
+var ErrNilValue = errors.New("record value should not be nil")
+
+func NewWriter(manager committed.PartManager, breakRatio int, treePath, partPath committed.Namespace) Writer {
+	return &writer{
+		manager:    manager,
+		breakRatio: breakRatio,
+		partPath:   partPath,
+		treePath:   treePath,
+	}
+}
+
+// WriteRecord writes a record to the current part, decides if should close part
+func (w *writer) WriteRecord(record graveler.ValueRecord) error {
+	if w.lastKey != nil && bytes.Compare(record.Key, w.lastKey) < 0 {
+		return ErrUnsorted
+	}
+	if record.Value == nil {
+		return ErrNilValue
+	}
+
+	var err error
+	if w.partWriter == nil {
+		w.partWriter, err = w.manager.GetWriter(w.partPath)
+		if err != nil {
+			return err
+		}
+	}
+
+	v, err := MarshalValue(record.Value)
+	if err != nil {
+		return err
+	}
+	err = w.partWriter.WriteRecord(committed.Record{Key: committed.Key(record.Key), Value: v})
+	if err != nil {
+		return err
+	}
+	w.lastKey = committed.Key(record.Key)
+	breakPoint, err := w.breakPoint(record.Key)
+	if err != nil {
+		return err
+	}
+	if breakPoint {
+		if err = w.closeCurrentPart(true); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (w *writer) closeCurrentPart(reachedBrakePoint bool) error {
+	if w.partWriter == nil {
+		return nil
+	}
+	wr, err := w.partWriter.Close()
+	if err != nil {
+		return err
+	}
+	part := Part{
+		ID:                wr.PartID,
+		MinKey:            wr.First,
+		MaxKey:            wr.Last,
+		EstimatedSize:     wr.Count,
+		ReachedBrakePoint: reachedBrakePoint,
+	}
+	err = w.writePartToTree(part)
+	if err != nil {
+		return err
+	}
+	w.partWriter = nil
+	return nil
+}
+
+func (w *writer) AddPart(part Part) error {
+	if w.lastKey != nil && bytes.Compare(part.MinKey, w.lastKey) < 0 {
+		return ErrUnsorted
+	}
+	if err := w.closeCurrentPart(false); err != nil {
+		return err
+	}
+	if err := w.writePartToTree(part); err != nil {
+		return err
+	}
+	w.lastKey = part.MaxKey
+	return nil
+}
+
+func (w *writer) SaveTree() (*graveler.TreeID, error) {
+	if err := w.closeCurrentPart(false); err != nil {
+		return nil, err
+	}
+	tree, err := w.treeWriter.Close()
+	if err != nil {
+		return nil, err
+	}
+	treeID := graveler.TreeID(tree.PartID)
+	return &treeID, nil
+}
+
+func hashToInt(s []byte) (uint32, error) {
+	h := fnv.New32a()
```

Comment:
A 32-bit hash makes me a bit worried.  The 64-bit version should just about be as efficient, and should behave better.  With 32 bits, the birthday paradox will ensure nastiness with fewer than 10_000 elements, and 50% chance of a collision with 2^16==65536 elements.  Rather than analyze, add bits: I hear 64-bit microprocessors are already available!